cmd/maya-exporter/app/collector/jivacollector.go

```diff
@@ -0,0 +1,129 @@
+package collector
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/openebs/maya/types/v1"
+)
+
+// NewJivaStatsExporter returns Jiva volume controller URL along with Path.
+func NewJivaStatsExporter(volumeControllerURL *url.URL) *VolumeExporter {
+	volumeControllerURL.Path = "/v1/stats"
+	return &VolumeExporter{
+		VolumeControllerURL: volumeControllerURL.String(),
+	}
+}
+
+// getVolumeStats is used to decode the response from the Jiva controller
+// response received by the client is in json format which then decoded
+// and mapped to VolumeMetrics.
+func (e *VolumeExporter) getVolumeStats(obj interface{}) error {
+	httpClient := http.DefaultClient
+	httpClient.Timeout = 1 * time.Second
+	resp, err := httpClient.Get(e.VolumeControllerURL)
+
+	if err != nil {
+		glog.Error("could not retrieve OpenEBS Volume controller metrics: %v", err)
+		return err
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(obj)
+
+	if err != nil {
+		glog.Error("could not decode OpenEBS Volume controller metrics: %v", err)
+		return err
+	}
+
+	defer resp.Body.Close()
+	return nil
+}
+
+// collect is used to set the values gathered from OpenEBS volume controller
+// to prometheus gauges.
+func (e *VolumeExporter) collect() error {
+	var (
+		metrics1 v1.VolumeMetrics
+		metrics2 v1.VolumeMetrics
+	)
+
+	err := e.getVolumeStats(&metrics1)
+	if err != nil {
+		glog.Error("Could not decode: %v", err)
+		return err
+	}
+
+	time.Sleep(1 * time.Second)
+	err = e.getVolumeStats(&metrics2)
+	if err != nil {
+		glog.Error("Could not decode: %v", err)
+		return err
+	}
+
+	// i and f is used for initial and final
+	rIOPS, _ := v1.ParseAndSubstract(metrics1.ReadIOPS, metrics2.ReadIOPS)
+	readIOPS.Set(rIOPS)
```

Comment:
Will it be better to have these counters, gauges etc as properties of volume exporter struct. These can registered in a map or something. Finally only the registered counters gauges etc are set. While setting them we can make use of MetricsDiff object to extract the exact property that is relevant to the counter.

Reply:
structure gets registered not the variables, These variables are not tied with any struct because of two reasons-:
1. while calling or passing the nil value to any associated methods, it may panic.
2. cstor doesn't exposes the other statistics (such as latency, uptime, readblockcount, writeblockcount etc.)
So let it be there as global variables just like other counters have been used.